### ISSUE TITLE
Canvas isn't refreshing on create.html and team.html on some computers?

### ISSUE THREAD

**Issue by hallpell** [stevekrouse/WoofJS#605]
I've seen this on two different student computers, so making an issue. I'm between somewhat and quite frustrated at trying to track this down, as I don't really have a good sense of steps to reproduce the issue.

When the issue is present:
- It looks like the game does 1-2 actions of forever loops, and then stops refreshing.
- The preview window has scroll bars on it that only show up when this problem is present
- There is (usually? I only verified this with one of the students) a second copy of the entire canvas (including another visible copy of sprites that have been made) that you can scroll down and see
- There was no evidence of issues in the console, though I might want to inspect the HTML a bit more if I can reproduce this

The issue is not consistently happening, so if you refresh the preview window it will sometimes work and sometimes not (feels like it's broken 80% of the time on the student I had more time to explore this issue with). The work-around is that this doesn't seem to happen in full.html, so we can have students always view their projects in full screen, which disrupts the clean workflow that is one of Woof's strongest aspects, but is at least functional.

The two examples I have were both on Windows machines, Chrome.

**Comment by madduccino** [comment]
Strangely enough I just saw this for the first time last week when someone was using Firefox - where there are two different iframes basically, one on top of each other. Is this something you've seen only recently?

**Comment by hallpell** [comment]
I think I've seen it in the past (over a year ago?), though I didn't understand it much at the time. I've had students that had issues with the Preview window which was then fixed when they went to full screen, which in retrospect probably shares root causes with this. I didn't notice the two iframes before, but I wasn't looking for it before. I think I at least have some loose threads to pull on once I have time to look into this. (Comparing create.html vs team.html, how we set up the window size for it, looking into the HTML that creates iframes to hold the canvases)

**Comment by hallpell** [comment]
This has started to appear more and more with (mostly new?) students. (Though possibly it's only with new students because students that have this issue don't continue with us?) We've had some classes where randomly divided groups have 4 students who all experience this issue. I don't really have many more theories as to what could be causing this, but various speculation/theories:

- Could this issue be more likely to come up before an account is created (when we're in a not-logged-in state)? I hope not, but it feels correlated based on my observations. I can't come up with a convincing causal link based on the code, but that doesn't mean it can't exist (and I could find it after more consideration?).
- @madduccino mentioned that reloading the Preview window (which should be refreshing the iframe, possibly by deleting it and creating a new one?) sometimes fixes it and sometimes doesn't, so it's possible there's something going on with failing to delete the older iframe? Especially auspicious is the code around here: https://github.com/stevekrouse/WoofJS/blob/3bc5e86950335a2f2275ebd160a78451d56c4bb6/create.html#L1746
- It appears that this issue is coming up more and more recently. Is there something that recently changed with browsers/DOM that breaks code in this area? It's getting frequent enough that I think this needs to get dealt with, but I really need to be able to replicate it on a computer I can dig into to make much more progress.

**Comment by hallpell** [comment]
I think I got a reasonable handle on this in #608 . It consistently replicated on Firefox on my computer, where it was adding two copies of the canvases into the iframe. I believe the cause was that firefox handles load events differently when you create iframes. Around https://github.com/stevekrouse/WoofJS/blob/42a2c231df3ddf922587df68eb4c413f11d343c8/create.html#L1768 we explicitly fire a load event on the newly created iframe, which is an issue in firefox because it will do it for us, so we end up with 2 copies of the event firing. I initially fixed it by causing our explicit event not to fire if we're in firefox, but realized it's probably better to go into woof.js https://github.com/stevekrouse/WoofJS/blob/42a2c231df3ddf922587df68eb4c413f11d343c8/woof.js#L150 and remove the project if it already exists. I think this should deal with the underlying cause better, but is still slightly clunky.

I'm not positive this fixes all the cases that I've seen, as it's been tricky to track down replication conditions, and I've only been able to play around with one setup extensively. However, I think this does take care of this. I'm going to leave this issue open for now, but it's possible this is taken care of.
